openpaas-james/mailbox/encrypted/memory/src/main/java/com/linagora/openpaas/encrypted/InMemoryKeystoreManager.scala

```diff
@@ -1,42 +1,81 @@
+    val keyId = KeyId.generate()
+    validateUsername(username)
+      .fold(_ => create(username, keyId, payload),
+        _ => update(username, keyId, payload))
+  }
 
-  override def listPublicKeys(username: Username): Publisher[PublicKey] = ???
+  override def listPublicKeys(username: Username): Publisher[PublicKey] =
+    validateUsername(username)
+      .fold(_ => SFlux.empty,
+        _ => SFlux.fromIterable(keystore(username)))
 
-  override def retrieveKey(username: Username, id: KeyId): Publisher[PublicKey] = {
-    val x: SMono[Set[PublicKey]] = SMono.fromCallable(() => keystore(username))
+  override def retrieveKey(username: Username, keyId: KeyId): Publisher[PublicKey] = {
+    val userKeys: SMono[Set[PublicKey]] = SMono.fromCallable(() => keystore(username))
 
-    usernameExists(username)
-      .fold(e => SMono.raiseError(e),
-        _ => x.map(keys => keys.find(key => key.id.equals(id)))
+    validateUsername(username)
+      .fold(_ => SMono.empty,
+        _ => userKeys.map(keys => keys.find(key => key.id.equals(keyId)))
           .filter(option => option.isDefined)
           .map(option => option.get)
-          .switchIfEmpty(SMono.raiseError(new IllegalArgumentException(s"Cannot find key $id"))))
+          .switchIfEmpty(SMono.raiseError(new IllegalArgumentException(s"Cannot find key $keyId"))))
   }
 
-  override def delete(username: Username, id: KeyId): Publisher[Void] = ???
+  override def delete(username: Username, keyId: KeyId): Publisher[Void] =
+    validateUsername(username)
+      .map(_ => keystore(username))
+      .flatMap(keys => checkKeyId(keys, keyId)
+        .map(key => keystore.put(username, keys -- Set(key))))
+      .fold(e => SMono.raiseError(e), _ => SMono.empty)
 
-  override def deleteAll(username: Username): Publisher[Void] = ???
+  private def checkKeyId(keys: Set[PublicKey], keyId: KeyId): Either[IllegalArgumentException, PublicKey] = {
+    if (keys.exists(key => key.id.equals(keyId))) {
+      Right(keys.find(k => k.id.equals(keyId)).get)
+    } else {
+      Left(new IllegalArgumentException(s"Could not find key $keyId"))
+    }
+  }
 
-  private def usernameExists(username: Username): Either[IllegalArgumentException, Boolean] = {
+  override def deleteAll(username: Username): Publisher[Void] =
+    validateUsername(username)
+      .fold(e => SMono.raiseError(e),
+        _ => SMono.fromCallable(() => keystore.remove(username)).`then`())
+
+  private def create(username: Username, keyId: KeyId, payload: Array[Byte]): Publisher[KeyId] =
+    SMono.fromCallable(() => keystore.put(username, Set(PublicKey(keyId, payload))))
+      .map[KeyId](_ => keyId)
+
+  private def update(username: Username, keyId: KeyId, payload: Array[Byte]): Publisher[KeyId] =
+    validatePayload(username, payload)
+      .fold(_ => updateWhenPayloadNotExist(username, keyId, payload),
+        existingKeyId => SMono.just(existingKeyId))
```

Comment:
You are doing a read-before your write pattern to ensure payload unicity.

I prefer to solve it by design by having `keyId=base64(sha-256(payload))`, and just do a dumb upsert.

It will also improve the Cassandra design / perfs.